src/artifact_providers/base.ts

```diff
@@ -240,23 +240,23 @@ export abstract class BaseArtifactProvider {
   ): Promise<RemoteArtifact[]> {
     this.logger.debug(`Fetching artifact list for revision \`${revision}\`.`);
     // check the cache first
-    const cached = this.fileListCache[revision];
-    if (cached) {
+    if (this.fileListCache[revision]) {
       this.logger.debug(`Found list in cache.`);
-      return cached;
+    } else {
+      // Cache the promise immediately to cause any subsequent calls during the
+      // fetch to use the pending promise instead of fetching again in parallel
```

Comment:
This change was required as now there can be concurrent calls to this method. I'd argue it should have been this way from the start though :)